React.Component API reference translation (#49)
reactjs/hu.reactjs.org: content/docs/reference-react-component.md

A React konstruktorokat általában kétféle okból használjuk:

* Inicializálni a [helyi állapotot](/docs/state-and-lifecycle.html) egy objektum `this.state`-hez rendelésével.
* Hozzákötni [eseménykezelő](/docs/handling-events.html) metódusokat egy komponenspéldányhoz.
  
**Nem szabad a `setState()`-et** meghívni a `constructor()`-ban. Ehelyett, ha a komponensednek helyi állapotot kell használnia, **rendeld hozzá a kezdő állapotot a `this.state`-hez** közvetlenül a konstruktorban:

```js
constructor(props) {
  super(props);
  // Ne hívd meg a this.setState()-et itt!
  this.state = { counter: 0 };
  this.handleClick = this.handleClick.bind(this);
}
```

A konstruktor az egyetlen hely, ahol a `this.state`-hez közvetlenül hozzárendelhetsz egy értéket. Mindenhol máshol a `this.setState()`-et használd ehelyett.

Kerüld el a mellékhatások bevezetését vagy a feliratkozásokat a konstruktorban. Ezekre az esetekre inkább a `componentDidMount()`-ot használd.

>Megjegyzés
>
>**Kerüld el a propok helyi állapotba való másolását! Ez egy gyakori hiba:**
>
>```js
>constructor(props) {
>  super(props);
>  // Ezt ne csináld!
>  this.state = { color: props.color };
>}
>```
>

> >A probléma az, hogy ez mind felesleges (a `this.props.color`-t közvetlenül is tudod használni ehelyett), mind pedig hibákat generál (a `color` prop frissülése nem fog megjelenni a helyi állapotban).

Reviewer: ```suggestion
>A probléma az, hogy ez mind felesleges (a `this.props.color`-t közvetlenül is tudod használni ehelyett), mint pedig hibákat generál (a `color` prop frissülése nem fog megjelenni a helyi állapotban).
```

Author: Sorry it's not a typo, it was meant to be 'mind ..., mind' as in 'both ..., and'
Feel free to suggest a better way of saying this :)

Ez egy jó alkalom arra, hogy változtatásokat hajts végre a DOM-on, amikor egy komponens frissült. Ez egy jó hely arra is, hogy hálózati kéréseket hajts végre, de csak akkor, ha összehasonlítod a jelenlegi propokat az előző propokkal (például nincs szükség a hálózati kérésre, ha a propok nem változtak). 

```js
componentDidUpdate(prevProps) {
  // Tipikus használat (ne felejtsd el a propokat összehasonlítani):
  if (this.props.userID !== prevProps.userID) {
    this.fetchData(this.props.userID);
  }
}
```

A **`setState()`-et akár azonnal is meghívhatod** a `componentDidUpdate()`-ben, de figyelj arra, hogy ennek **egy feltételben kell elhelyezkednie**, mint a fenti példában, különben egy végtelen ciklus lesz az eredmény. Ez ugyancsak okoz egy extra újrarenderelést, ami bár nem látható a felhasználó számára, de befolyásolhatja a komponens teljesítményét. Ha a helyi állapot egy részét a propból képzed le, fontold meg ehelyett a prop közvetlen használatát. Itt olvashatsz többet a [hibákról, amik a propok helyi állapotba való másolásakor fordulnak elő](/blog/2018/06/07/you-probably-dont-need-derived-state.html).

> Ha a komponensed implementálja a `getSnapshotBeforeUpdate()` életciklus metódust (ami ritka), az érték, amit visszaad, `componentDidUpdate()` egy harmadik paramétereként lesz átadva. Egyébként ez a paraméter undefined értékű lesz. 

Reviewer: ```suggestion
Ha a komponensed implementálja a `getSnapshotBeforeUpdate()` életciklus metódust (ami ritka), az érték, amit visszaad, a `componentDidUpdate()` egy harmadik paramétereként lesz átadva. Egyébként ez a paraméter undefined értékű lesz. 
```

static getDerivedStateFromProps(props, state)
```

> A `getDerivedStateFromProps` közvetlenül a renderelés előtt hajtódik végre a kezdeti renderelésnél és frissítéskor is. Ennek egy objektumot kell visszaadnia az állapot módosításához, vagy null-t, hogy ne módosítson semmit..

Reviewer: ```suggestion
A `getDerivedStateFromProps` közvetlenül a renderelés előtt hajtódik végre a kezdeti renderelésnél és frissítéskor is. Ennek egy objektumot kell visszaadnia az állapot módosításához, vagy null-t, hogy ne módosítson semmit.
```

Ez a metódus csak [ritka használati esetekre](/blog/2018/06/07/you-probably-dont-need-derived-state.html#when-to-use-derived-state) való, amikor a helyi állapot a propok változásaitól függ. Például hasznos lehet egy `<Transition>` komponens implementálására, ami összehasonlítja az előző és a következő gyermekeit, hogy eldöntse, melyiket animálja kifelé vagy befelé.

A leszármaztatott állapot beszédes kódhoz vezet, ami a komponenseidet kevésbé átláthatóvá teszi.  
[Győződj meg róla, hogy tisztában vagy egyszerűbb alternatívákkal is:](/blog/2018/06/07/you-probably-dont-need-derived-state.html)

> * Ha muszály **mellékhatást végrehajtanod** (például adatlekérés vagy animáció) egy propokban történt változás hatására, használd a [`componentDidUpdate`](#componentdidupdate) életciklus metódust ehelyett.

Reviewer: ```suggestion
* Ha muszáj **mellékhatást végrehajtanod** (például adatlekérés vagy animáció) egy propokban történt változás hatására, használd a [`componentDidUpdate`](#componentdidupdate) életciklus metódust ehelyett.
```

http://muszaj.hu/ :slightly_smiling_face:

* Ha **adatokat akarsz újraszámítani csak akkor, ha prop változás történt**, [használj egy memoizálás segítőt](/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization).

* Ha **"újra kell inicializálni" az állapot egy részét prop változás hatására**, fontold meg a komponens [teljesen kontrollálttá tételét](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-controlled-component), vagy [teljesen kontrollálatlaná tételét egy `key`-vel](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key) ehelyett.

> Ennek a metódusnak nincs hozzáférése a komponenspéldányhoz. Ha szeretnéd, újrahasználhatsz némi kódot a `getDerivedStateFromProps()` és a többi osztálymetódus között a propok és helyi állapot tiszta függvényeinek kiemelésével az osztálydefiníción kivülre.

Reviewer: ```suggestion
Ennek a metódusnak nincs hozzáférése a komponenspéldányhoz. Ha szeretnéd, újrahasználhatsz némi kódot a `getDerivedStateFromProps()` és a többi osztálymetódus között a propok és helyi állapot tiszta függvényeinek kiemelésével az osztálydefiníción kívülre.
```

getSnapshotBeforeUpdate(prevProps, prevState)
```

> A `getSnapshotBeforeUpdate()` rögtön azelőtt hívódik meg, hogy az legutóbb renderelt kimenet el van mentve pl. a DOM-ba. Ez lehetővé teszi, hogy a komponens kiragadjon információkat a DOM-ból (pl. görgetési pozíció), mielőtt az potenciálisan megváltozna. Bármely érték, amit ez a metódus visszaad, a `componentDidUpdate()`-nek lesz átadva.

Reviewer: ```suggestion
A `getSnapshotBeforeUpdate()` rögtön azelőtt hívódik meg, hogy a legutóbb renderelt kimenet el van mentve pl. a DOM-ba. Ez lehetővé teszi, hogy a komponens kiragadjon információkat a DOM-ból (pl. görgetési pozíció), mielőtt az potenciálisan megváltozna. Bármely érték, amit ez a metódus visszaad, a `componentDidUpdate()`-nek lesz átadva.
```

> Ez a használati eset nem gyakori, de előfordulhat olyan felhasználó felületeken, mind például egy csetalkalmazás, aminek speciálisan kell kezelnie a görgetési pozíciót.

Reviewer: ```suggestion
Ez a használati eset nem gyakori, de előfordulhat olyan felhasználói felületeken, mint például egy csevegőalkalmazás, aminek speciálisan kell kezelnie a görgetési pozíciót.
```

Do not think we use "cset" in Hungarian

Egy pillanatfelvételi érték (vagy `null`) legyen a visszatérés értéke.

Például:

`embed:react-component-reference/get-snapshot-before-update.js`

A fenti példákban fontos, hogy kiolvassuk a `scrollHeight` értékét a `getSnapshotBeforeUpdate`-ben, mivel időeltolódás lehet a "renderelési" életciklus fázis (mint a `render`) és "elmentési" életciklus fázis (mint a `getSnapshotBeforeUpdate` vagy a `componentDidUpdate`) között.

* * *

### Hibahatárok {#error-boundaries}

> A [hibahatárok](/docs/error-boundaries.html) React komponensek, amik elkapják a JavaScript hibákat bárhol a komponensfában, kinaplózzák ezeket, és egy tartalék UI-t jelenítenek meg az összeomlott komponensfa helyett. A hibahatárok elkapják a hibábak renderelés közben, az életciklusmetódusokban, és a teljes gyermekkomponensfa konstruktoraiban.

Reviewer: ```suggestion
A [hibahatárok](/docs/error-boundaries.html) React komponensek, amik elkapják a JavaScript hibákat bárhol a komponensfában, kinaplózzák ezeket, és egy tartalék UI-t jelenítenek meg az összeomlott komponensfa helyett. A hibahatárok elkapják a hibákat renderelés közben, az életciklusmetódusokban, és a teljes gyermekkomponensfa konstruktoraiban.
```

UNSAFE_componentWillReceiveProps(nextProps)
```

> Megjegyzés
>

> > Ezt az életciklust előzőleg `componentWillReceiveProps`-nek hívták. Ez a név a 17-es verzióig működni fog. Használd a [`rename-unsafe-lifecycles` kódmódosítót](https://github.com/reactjs/react-codemod#rename-unsafe-lifecycles), hogy automatikusan frissítsd a komponenseidet.

Reviewer: ```suggestion
> Ezt az életciklust előzőleg `componentWillReceiveProps`-nak hívták. Ez a név a 17-es verzióig működni fog. Használd a [`rename-unsafe-lifecycles` kódmódosítót](https://github.com/reactjs/react-codemod#rename-unsafe-lifecycles), hogy automatikusan frissítsd a komponenseidet.
```

> Megjegyzés:
>
> Ennek az életciklusmetódusnak a használata gyakran vezet hibákhoz és inkonzisztenciákhoz
>
> * Ha **mellékhatást kell végrehajtanod**, (például adatlekérés vagy egy animáció) a propokban történt változások határásra, használd a [`componentDidUpdate`](#componentdidupdate) életciklust ehelyett.
> * Ha a `componentWillReceiveProps`-ot használtad az **adatok újraszámítására a propok változásai alapján**, [használj memoizálás segítőt ehelyett](/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization).

> > * Ha a `componentWillReceiveProps`-ot használtad, hogy **"újrainicializáld" az  állapot egy részét a propok változásának hatására**, fontold meg a komponense [teljesen kontrolláltá tételét](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-controlled-component), vagy [teljesen kontrollálatlanná egy `key`-vel](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key) ehelyett.

Reviewer: ```suggestion
> * Ha a `componentWillReceiveProps`-ot használtad, hogy **"újrainicializáld" az  állapot egy részét a propok változásának hatására**, fontold meg a komponens [teljesen kontrollálttá tételét](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-controlled-component), vagy [teljesen kontrollálatlanná egy `key`-vel](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key) ehelyett.
```

>
> A többi használati esetben [kövesd az ajánlásokat, amik ebben a származtatott állapotról szóló blogbejegyzésben találhatóak](/blog/2018/06/07/you-probably-dont-need-derived-state.html).

> Az `UNSAFE_componentWillReceiveProps()` azelőtt hívódik meg, mielőtt egy betöltött komponens új propokat kapna. Ha meg kell változtatnod az állapotot a propok változásának hatására, (például, hogy újrainicializálás), összehasonlíthatod a `this.props`-ot a `nextProps`-szal és állapotot változtathatsz a `this.setState()`-tel ebben a metódusban.

Reviewer: ```suggestion
Az `UNSAFE_componentWillReceiveProps()` azelőtt hívódik meg, mielőtt egy betöltött komponens új propokat kapna. Ha meg kell változtatnod az állapotot a propok változásának hatására, (például, újrainicializálás), összehasonlíthatod a `this.props`-ot a `nextProps`-szal és állapotot változtathatsz a `this.setState()`-tel ebben a metódusban.
```

Megjegyzendő, hogy ha egy szülő komponens miatt renderelődik újra a komponensed, ez a metódus meghívódik akkor is, ha a propok nem változtak. Hasonlítsd össze a jelenlegi és következő értékeket, ha csak a változásokat akarod lekezelni.

> A React nem hívja meg a `UNSAFE_componentWillReceiveProps()`-t a kezdeti prop értékekkel a [betöltés](#mounting) közben. Csak akkor hívja meg ezt a metódust, ha a komponens néhány propja frissülhet. A `this.setState()` meghívása általában nem idézi elő a `UNSAFE_componentWillReceiveProps()` meghívását.

Reviewer: ```suggestion
A React nem hívja meg az `UNSAFE_componentWillReceiveProps()`-t a kezdeti prop értékekkel a [betöltés](#mounting) közben. Csak akkor hívja meg ezt a metódust, ha a komponens néhány propja frissülhet. A `this.setState()` meghívása általában nem idézi elő az `UNSAFE_componentWillReceiveProps()` meghívását.
```

UNSAFE_componentWillUpdate(nextProps, nextState)
```

> Mejegyzés
>
> Ezt az életciklust előzőleg `componentWillUpdate`-nek hívták. Ez a név a 17-es verzióig működni fog. Használd a [`rename-unsafe-lifecycles` kódmódosítót](https://github.com/reactjs/react-codemod#rename-unsafe-lifecycles), hogy automatikusan frissítsd a komponenseidet.

> A `UNSAFE_componentWillUpdate()` közvetlenül renderelés előtt hívódik meg, amikor új propok és állapotok kerülnek fogadásra. Ez egy jó lehetőség arra, hogy előkészítést végezzünk el egy frissítés előtt. Ez a metódus nem hívódik meg a kezdeti rendereléskor.

Reviewer: ```suggestion
Az `UNSAFE_componentWillUpdate()` közvetlenül renderelés előtt hívódik meg, amikor új propok és állapotok kerülnek fogadásra. Ez egy jó lehetőség arra, hogy előkészítést végezzünk el egy frissítés előtt. Ez a metódus nem hívódik meg a kezdeti rendereléskor.
```

Megjegyzendő, hogy itt nem hívhatod meg a `this.setState()`-et; valamint bármi mást sem (pl. kezdeményezni egy Redux műveletet), ami egy frissítést kezdeményezne a React komponensben a `UNSAFE_componentWillUpdate()` visszatérése előtt.

Általában ez a metódus kicserélhető a `componentDidUpdate()`-tel. Ha ebben a metódusban a DOM-ból olvastál ki (pl. a görgetési pozíció elmentése), ezt a logikát áthelyezheted a `getSnapshotBeforeUpdate()`-be.

> Megjegyzés
>
> A `UNSAFE_componentWillUpdate()` nem hívódik meg, ha a [`shouldComponentUpdate()`](#shouldcomponentupdate) hamisat ad vissza.

* * *

## Egyéb API-k {#other-apis-1}

A fenti életciklusmetódusokkal ellentétben (amiket a React hív meg neked), a következő metódusokat *te magad* hívhatod meg a komponenseidből.

Csak ez a kettő van: `setState()` és `forceUpdate()`.

### `setState()` {#setstate}

```javascript
setState(updater[, callback])
```

>  A `setState()` egy várakozási sorba helyezi a komponens állapotának változásait és utasítja a Reactet, hogy ez a komponens a gyermekeivel együtt újrarendereljen az új állapottal. Ez az elsődleges metódus a kezelői felület frissítéséhez a eseménykezelők és szerver általi válasz hatására.

Reviewer: ```suggestion
 A `setState()` egy várakozási sorba helyezi a komponens állapotának változásait és utasítja a Reactet, hogy ez a komponens a gyermekeivel együtt újrarendereljen az új állapottal. Ez az elsődleges metódus a kezelői felület frissítéséhez az eseménykezelők és szerver általi válasz hatására.
```

> Gondolj a `setState()`-re mint *kérés*, egy azonnali parancs helyett a komponens frissítésére. A jobb észlelt teljesítmény miatt a React dönthet úgy, hogy késlelteti a végrehajtást, és aztán több komponenst is egyszerre frissít. A React nem garantálja, hogy az állapotfrissítések azonnal megtörténnek.

Reviewer: ```suggestion
Gondolj a `setState()`-re mint *kérés*, egy azonnali parancs helyett a komponens frissítésére. A jobb észlelt teljesítmény érdekében a React dönthet úgy, hogy késlelteti a végrehajtást, és aztán több komponenst is egyszerre frissít. A React nem garantálja, hogy az állapotfrissítések azonnal megtörténnek.
```

> A `setState()` nem mindig frissíti rögtön a komponenst. Lehet, hogy összevonja más frissítésekkel, vagy késlelteti a frissítést. Emiatt a `this.state`-t kiolvasása rögtön a `setState()` meghívása után egy potenciális buktató. Ehelyett használd a `componentDidUpdate`-t vagy a `setState` callbackjét (`setState(updater, callback)`), ezek közül bármelyik garantáltan lefut egy frissítés megtörténte után . Ha az állapotot egy előző állapot alapján kell beállítanod, olvass az `updater` argumentumról alább.

Reviewer: ```suggestion
A `setState()` nem mindig frissíti rögtön a komponenst. Lehet, hogy összevonja más frissítésekkel, vagy késlelteti a frissítést. Emiatt a `this.state` kiolvasása rögtön a `setState()` meghívása után egy potenciális buktató. Ehelyett használd a `componentDidUpdate`-t vagy a `setState` callbackjét (`setState(updater, callback)`), ezek közül bármelyik garantáltan lefut egy frissítés megtörténte után. Ha az állapotot egy előző állapot alapján kell beállítanod, olvass az `updater` argumentumról alább.
```

> A `setState()` mindig újrarendereléshez vezet, kivéve, ha a `shouldComponentUpdate()` `false`-ot ad vissza. Ha változtatható objektumokat használsz és nem lehet implementálni feltételes renderelő logikát a `shouldComponentUpdate()`-ben, a felesleges renderelések elkerülése érdekében a `setState()`-et csak akkor hívd meg, ha az új állapot különbözik az előzőtől.

Reviewer: ```suggestion
A `setState()` mindig újrarendereléshez vezet, kivéve, ha a `shouldComponentUpdate()` `false`-ot ad vissza. Ha megváltoztatható objektumokat használsz és nem lehet implementálni feltételes renderelő logikát a `shouldComponentUpdate()`-ben, a felesleges renderelések elkerülése érdekében a `setState()`-et csak akkor hívd meg, ha az új állapot különbözik az előzőtől.
```

Only for consistency with existing translations.

Az első argumentum egy `updater` függvény az alábbi szignatúrával:

```javascript
(state, props) => stateChange
```

A `state` egy referencia a komponens állapotára a frissítés idejében. Ezt nem szabad közvetlenül módosítani. Ehelyett a változásokat reprezentáld egy új objektumban a `state` és `props` alapján. Például tegyük fel, hogy meg akarjuk növelni egy állapotváltozó értéket `props.step`-pel:

```javascript
this.setState((state, props) => {
  return {counter: state.counter + props.step};
});
```

Mind `state` és `props` értéke, amit a frissítőfüggvény (updater) által kaptunk, garantáltan friss lesz. A frissítőfüggvény kimenete a `state`-tel kerül sekélyen összefésülésre.

A `setState()` második paramétere egy opcionális callback függvény, ami akkor lesz meghívva, amikor a `setState` lefutott és a komponens újrarenderelt. Mi általában hasonló logikához a `componentDidUpdate()`-et ajánljuk ehelyett.

A `setState()` első argumentuma függvény helyett lehet objektum is:

```javascript
setState(stateChange[, callback])
```

> Ez a `stateChange` egy sekély összefésülését végzi el az új állapotba, pl., amikor egy bevásárlókocsi elemeinek számát változtatod meg:

Reviewer: ```suggestion
Ez a `stateChange` egy sekély összefésülését végzi el az új állapotba, pl., amikor egy kosár elemeinek számát változtatod meg:
```

On the web, I think we use "kosár", instead of "bevásárlókocsi". If I am wrong, ignore this.

component.forceUpdate(callback)
```

Alapesetben, amikor a komponensed állapota vagy propjai változnak, a komponensed újrarenderelődik. Ha a `render()` metódusod valami más adattól is függ, utasíthatod a Reactet az újrarenderelésre a `forceUpdate()` meghívásával.

> A `forceUpdate()` meghívása a `render()` hívását fogja előidézni a komponensben, kihagyva a `shouldComponentUpdate()`-t. Ez ugyanakkor előidézi a normális életciklusmetódusok meghívását a gyermekkomponensekben a `shouldComponentUpdate()`-tel együtt. Viszont a React ugyanúgy csak akkor frissíti a DOM-ot, ha a változás történt.

Reviewer: ```suggestion
A `forceUpdate()` meghívása a `render()` hívását fogja előidézni a komponensben, kihagyva a `shouldComponentUpdate()`-t. Ez ugyanakkor előidézi a normális életciklusmetódusok meghívását a gyermekkomponensekben a `shouldComponentUpdate()`-tel együtt. Viszont a React ugyanúgy csak akkor frissíti a DOM-ot, ha változás történt.
```

Általában elkerülendő a `forceUpdate()` használata és csak a `this.props` és `this.state`-ből olvass a `render()`-ben.

* * *

## Osztálytulajdonságok {#class-properties-1}

### `defaultProps` {#defaultprops}

> A `defaultProps` egy tulajdonságként lehet definiálva magán a komponensen, hogy beállítsa a propok alapértékeit az osztályon. Ez az undefined értékű propoknál használatos, de nem a null értékűekre. Például:

Reviewer: ```suggestion
A `defaultProps` egy tulajdonságként lehet definiálva magán a komponensen, hogy beállítsa a propok alapértékeit az osztályon. Ez az undefined értékű propoknál használatos, a null értékűeknél nem. Például:
```
